libuavcan/include/libuavcan/media/interfaces.hpp

```diff
@@ -0,0 +1,255 @@
+     * @param  out_frames_written
+     *                      The number of frames written. If this is less than frames_len then frames
+     *                      [0 - out_frames_written) were enqueued for transmission. Frames
+     *                      [out_frames_written - frames_len) were not able to be sent. Nominally this is
+     *                      due to the internal queues being full.
+     * @return libuavcan::Result::success_partial if some but not all of the frames were written.
+     *         libuavcan::Result::buffer_full if no frames could be written because the buffer was
+     *         full.
+     *         libuavcan::Result::success if all frames were written.
+     */
+    virtual libuavcan::Result write(const FrameT (&frames)[MaxTxFrames],
+                                    std::size_t  frames_len,
+                                    std::size_t& out_frames_written) = 0;
+
+    /**
+     * Non-blocking reception.
+     *
+     * Timestamps should be provided by the media driver, ideally by the hardware peripheral.
+     *
+     * While the monotonic timestamp is required the UAVCAN protoocol can tolerate imprecision since
+     * these timestamps are used only for protocol timing validation  (i.e. transfer timeouts and
+     * inter-transfer intervals).
+     *
+     * @return libuavcan::Result::success If no errors occurred.
+     */
+    virtual libuavcan::Result read(FrameT (&out_frames)[MaxRxFrames], std::size_t& out_frames_read) = 0;
+};
+
+/**
+ * Manages access to and the lifecycle of media interfaces to a single bus type. A given InterfaceManager
+ * shall only expose interfaces to busses that are used as redundant media for the same transport (i.e. An
+ * interface manager should define a single logical bus). How manager objects are exposed to an application
+ * is not specified by libuavcan.
+ *
+ * @tparam  InterfaceT              The type to use for interfaces. Must implement Interface.
+ * @tparam  InterfacePtrT           The pointer type to use for InterfaceT pointers in this class. On
+ *                                  systems without dynamic memory this can probably be simply <code>InterfaceT*</code>
+ *                                  but where dynamic memory is used some sort of smart pointer is recommended.
+ * @tparam  MaxSelectInterfacesVal  The maximum number of interfaces the manager's select method can support.
+ *                                  This is required since the media layer interfaces do not require dynamic memory.
+ *                                  It's also not expected that more than two or three interfaces would ever be
+ *                                  required (i.e. it is expected that the manager's select will be used to wait
+ *                                  on a redundant set of interfaces to a single bus).
+ */
+template <typename InterfaceT, typename InterfacePtrT, std::size_t MaxSelectInterfacesVal = 2>
```

Comment:
I'll change to 3 but keep the default since this is something that an implementer can think about after the initial bring-up on their platform. The pointer type I want them to have to think about since it's an important design decision that should be made up-front.